Sources/NIOSOCKS/Channel Handlers/SOCKSClientHandler.swift

```diff
@@ -0,0 +1,196 @@
+            guard self.state.shouldBeginHandshake else {
+                return
+            }
+            self.handleAction(self.state.connectionEstablished(), context: context)
+        }
+    }
+    
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        
+        // if we've established the connection then forward on the data
+        guard !self.state.proxyEstablished else {
+            context.fireChannelRead(data)
+            return
+        }
+        
+        var buffer = self.unwrapInboundIn(data)
+        self.buffered.writeBuffer(&buffer)
+        do {
+            let action = try self.state.receiveBuffer(&self.buffered)
+            switch action {
+            case .waitForMoreData:
+                break // do nothing, we've buffered the data already
+            default:
+                self.handleAction(action, context: context)
+            }
+        } catch {
+            context.fireErrorCaught(error)
+            context.close(mode: .all, promise: nil)
+        }
+    }
+    
+    public func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        guard self.state.proxyEstablished else {
+            self.bufferedWrites.append((data, promise))
+            return
+        }
+        context.write(data, promise: nil)
+    }
+    
+}
+
+extension SOCKSClientHandler {
+    
+    func startHandshake(context: ChannelHandlerContext) {
+        
+        // if the handshake has already begun
+        // or completed, then don't start it again
+        guard self.state.shouldBeginHandshake else {
+            return
+        }
+        
+        let greeting = ClientGreeting(
+            methods: self.supportedAuthenticationMethods
+        )
+        let capacity = 2 + self.supportedAuthenticationMethods.count
+        var buffer = context.channel.allocator.buffer(capacity: capacity)
+        buffer.writeClientGreeting(greeting)
+        
+        do {
+            try self.state.sendClientGreeting(greeting)
+        } catch {
+            context.fireErrorCaught(error)
+            return
+        }
+        
+        context.writeAndFlush(self.wrapOutboundOut(buffer), promise: nil)
+        context.fireChannelActive()
+    }
+    
+    func handleAction(_ action: ClientAction, context: ChannelHandlerContext) {
+        do {
+            switch action {
+            case .sendGreeting:
+                self.startHandshake(context: context)
+            case .authenticateIfNeeded(let method):
+                try self.handleActionAuthenticateIfNeeded(method: method, context: context)
+            case .sendRequest:
+                self.handleActionSendRequest(context: context)
+            case .proxyEstablished:
+                self.handleActionProxyEstablished(context: context)
+            case .waitForMoreData:
+                break
+            }
+        } catch {
+            context.fireErrorCaught(error)
```

Comment:
I don't think so, the connection should close instead.

Reply:
Sure, but closure may not be instantaneous.

Reply:
Fair point. I've refactored so that the error is already known by the state machine, so we don't need to feed it back in, but the state machine is aware and won't progress. I'll add a test for this too.